# This script is used to reanalyze the Sade-Feldman et al data on intratumoral immune cells to investigate SNX9+ vs SNX9- 
# Generated by Marcel Trefny on 9.11.2022
# Insert path of script below at setwd()
# The GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt file has to be downloaded from GEO (due to size). 
# Then unzip this file and place into input/raw folder. Then delete the first empty column/row entry. Then begin script.
# There is an easy entry point starting directly with the CD8T cells (find readRDS(sce_CD8T.rds))

#install packages
BiocManager::install("SingleCellExperiment", ask = FALSE)
BiocManager::install("scater", ask = FALSE)
BiocManager::install("uwot", ask = FALSE)
BiocManager::install("scran", ask = FALSE)
BiocManager::install("Rtsne", ask = FALSE)
BiocManager::install("batchelor", ask = FALSE)
BiocManager::install("BiocSingular", ask = FALSE)
BiocManager::install("BiocNeighbors", ask = FALSE)
BiocManager::install("rsvd", ask = FALSE)
BiocManager::install("igraph", ask = FALSE)
BiocManager::install("rowr", ask = FALSE)
BiocManager::install("tidyr", ask = FALSE)
BiocManager::install("dplyr", ask = FALSE)
BiocManager::install("ggplot2", ask = FALSE)
BiocManager::install("Rtsne", ask = FALSE)

# load packages
library(batchelor)
library(BiocSingular)
library(BiocNeighbors)
library(Rtsne)
library(rsvd)
library(tidyr) 
library(dplyr)
library(igraph)
library(SingleCellExperiment)
library(scater)
library(scran)
library(Rtsne)
library(ggplot2)


########################################################################

#Data Loading
path <- #! INSERT PATH OF SCRIPT HERE

setwd(path)
set.seed(123)

# # This code is used to generate counts_dg.rds from original GEO dataset by formating the list
# # This code takes lots of memory and time
# # !The GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt file has to be downloaded from GEO (due to size). 
files_row_column <- read.table("raw/GSE120575_Sade_Feldman_melanoma_single_cells_TPM_GEO.txt" , sep = "\t", skip = 2,  check.names= FALSE, header = FALSE, stringsAsFactors = FALSE )
counts <- files_row_column[,-c(1)] #first column are gene names
counts <- files_row_column[,-c(1,16292)] #last column is an artefact from the file loading

files_column_data_header <- read.table("raw/single_cells_TPM_GEO_header.txt",  sep = "\t",  check.names= FALSE, header = FALSE, stringsAsFactors = FALSE )
files_column_data_header <- data.frame(t(files_column_data_header))
colnames(files_column_data_header) <- c("title", "patientID")
head(files_column_data_header)
dim(counts)
dim(files_column_data_header)
colnames(counts) <- files_column_data_header$title

counts_dg <- as(as.matrix(counts), "dgCMatrix")
saveRDS(counts_dg, file = "counts_dg.rds")


counts_dg <- readRDS("counts_dg.rds")
head(counts_dg)

dim(counts_dg) #55737 16291

files_column_data <- read.table("GSE120575_patient_ID_single_cells_formatted.txt",  sep = "\t",  check.names= FALSE, header = TRUE, stringsAsFactors = FALSE )
files_column_data <- files_column_data[,1:7]
colnames(files_column_data) <- c("sample", "title", "source", "organism", "patientID", "response", "therapy")
head(files_column_data)
dim(files_column_data) #16291     7

genes <- read.table("genes.txt",  sep = "\t",  check.names= FALSE, header = FALSE, stringsAsFactors = FALSE )
head(genes)

counts_dg[55730:55737,16285:16291] #the last column is NA, thus remove
counts_dg <- counts_dg[,-c(ncol(counts_dg))] 
files_column_data <- files_column_data[-c(nrow(files_column_data)),] #also remove this cell's information


sce <- SingleCellExperiment(assays=SimpleList(counts=counts_dg), rowData = data.frame("GeneSymbol" = genes$V1),
                            colData=files_column_data)
rownames(sce) <- make.names(rowData(sce)$GeneSymbol) #need to make_names because of genes that start with a number
sce

sce$timepoint <- gsub("_.*$","", colData(sce)$patientID)
sce$patient <- gsub("^.*_P","P", colData(sce)$patientID)
sce$timepoint <- factor(sce$timepoint, levels = c("Pre", "Post"))

dim(assay(sce, "counts"))
head(rowData(sce))
head(colData(sce))
nrow(counts(sce))
ncol(counts(sce))

mean(counts(sce) == 0)
class(counts(sce))
counts(sce)[1:10, 1:10]

## Total number of detected transcripts
nrow(counts(sce)) #55737
sum(rowSums(counts(sce)) > 0)

########################################################################################################
############# Filtering of genes and cells
########################################################################################################
protein_coding_genes <- read.table("ProteinCodingGenes.txt", header = TRUE, sep = "\t")
protein_coding_genes <- protein_coding_genes$GeneSymbol

#focus on protein coding genes as in publication
summary(rowData(sce)$GeneSymbol %in% protein_coding_genes)
# Mode   FALSE    TRUE 
# logical   30146   20367 
sce <- sce[ rowData(sce)$GeneSymbol %in% protein_coding_genes,] 

#exclude all transcripts which were not detected in any cell
sce <- sce[rowSums(counts(sce)) > 0,] #50513

## Genes detected in single cells
summary(colSums(counts(sce) > 0)) #mean 1997 genes per cell

#exclude non-immune cells as in paper, but only with CD45 as cutoff
non_immune <- counts(sce)[rowData(sce)$GeneSymbol == "PTPRC", ] == 0
sce <- sce[,!non_immune] 


housekeeping_genes <- read.table("HousekeepingGenes.txt", header = TRUE, sep = "\t")
housekeeping_genes <- housekeeping_genes$GeneSymbol

housekeeping_mean_per_cell <- colMeans( counts(sce)[rowData(sce)$GeneSymbol %in% housekeeping_genes,])
summary(housekeeping_mean_per_cell > 2.5) #True for almost all cells
plot(sort(housekeeping_mean_per_cell))
sce <- sce[,housekeeping_mean_per_cell > 2.5] # take only cells with more than 2.5 log2(TPM+1) as in paper

#exclude genes that were not expressed with log2TPM+1 of >4.5 in at least 10 cells
summary(rowSums(counts(sce) > 4.5)>10) 
# Mode   FALSE    TRUE 
# logical    4521   15846 
sce <- sce[rowSums(counts(sce) > 4.5)>10,]


####################################
#Extract QC parameters and plot

mt <- rownames(sce)[grep("^MT-", rowData(sce)$GeneSymbol)] #mitochondrial genes
mt

sce <- addPerFeatureQC(sce) #scores on all genes
sce <- addPerCellQC(sce, subsets = list(MT = mt)) #scores on mitochondrial genes

plotColData(sce, x = "sum", y="detected", colour_by="patientID") 
colnames(rowData(sce))
colnames(colData(sce))

plot(sort(sce$subsets_MT_percent),   main = "percent mitochondrial")

############################################################
#Filtering of low gene detection and high mitochondrial DNA = death contamination

# low_detected <- isOutlier(sce$detected, type = "lower", 
#                           log = TRUE, nmads = 4)
high_mt <- isOutlier(sce$subsets_MT_percent, type = "higher",
                     log = FALSE, nmads = 3) #cells that are dying -> 3 standard deviations from mean mitochondrial percentage away
duplets <- isOutlier(sce$sum, type = "higher",
                     log = FALSE, nmads = 4) #doublets = cells with lots of genes detected

#duplets = very high number of counts
plot(rank(-sce$sum), sce$sum, col = duplets + 1 )
# dying cells = high content of mtDNA
plot(rank(sce$subsets_MT_percent), sce$subsets_MT_percent,
     col = high_mt + 1)

#based on this cut away the unwanted cells
sce$retain <-  !high_mt & !duplets
table(sce$retain) 
## 
#FALSE  TRUE 
# 519 15390 

sce <- sce[, sce$retain] #delete cells that did not pass QC

dim(sce)
# 15846 15390



########################################################################
#### Normalization
#look at the different library sizes of each cell with some patients having much more transcripts per cell sequenced
plot(colSums(counts(sce)) )

cpms <- assay(sce, "counts")
libsizes <- colSums(cpms)
size.factors <- libsizes/mean(libsizes)
logcounts(sce) <- t(t(cpms)/size.factors) + 1
assayNames(sce)
plot(colSums(logcounts(sce)))


## Fit a trend
dec.trend <- modelGeneVar(sce)
fit.trend <- metadata(dec.trend)
plot(fit.trend$mean, fit.trend$var, xlab = "Mean of log-expression",
     ylab = "Variance of log-expression")
curve(fit.trend$trend(x), col = "dodgerblue", add = TRUE, lwd = 2)
head(dec.trend[order(dec.trend$bio, decreasing = TRUE), ])


sce <- runPCA(sce, exprs_values = "logcounts", ncomponents = 50, 
              ntop = 4000)
sce
reducedDimNames(sce)


pdf(file = "../plots/reducedDimensions_PCA_tSNE_UMAP.pdf", width = 8)

  plotReducedDim(sce, "PCA", colour_by = "detected")
  plotReducedDim(sce, "PCA", colour_by = "patientID")
  plotReducedDim(sce, "PCA", colour_by = "therapy")

dev.off()
#perform tSNE
sce <- runTSNE(sce, dimred = "PCA",  perplexity = 30)
reducedDimNames(sce)

sce <- runUMAP(sce, dimred = "PCA")
reducedDimNames(sce)

saveRDS(sce, file = "sce.rds")


sce <- readRDS("sce.rds")
infos <- c(   "therapy", "timepoint")
table(colData(sce)[,infos])

plotReducedDim(sce, "TSNE", colour_by = "TLR9")
plotReducedDim(sce, "TSNE", colour_by = "SNX9")
plotReducedDim(sce, "TSNE", colour_by = "TCF7")
plotReducedDim(sce, "TSNE", colour_by = "CD8A")
plotReducedDim(sce, "TSNE", colour_by = "CD4")

plotReducedDim(sce, "TSNE", colour_by = "subsets_MT_percent")
plotReducedDim(sce, "TSNE", colour_by = "sum")
plotReducedDim(sce, "TSNE", colour_by = "detected")
plotReducedDim(sce, "TSNE", colour_by = "response")
plotReducedDim(sce, "TSNE", colour_by = "patientID")
plotReducedDim(sce, "TSNE", colour_by = "therapy")


#recreate their cluster ids from paper supplentary file
cluster_ID_publication <- read.table("cluster_ID_Paper.txt", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
head(cluster_ID_publication)
summary(cluster_ID_publication$title %in% colData(sce)$title)

sce$clusterSadeFeldman <-  left_join(data.frame(colData(sce)), cluster_ID_publication, by = "title")$clusterSadeFeldman
sce$clusterSadeFeldman <- as.factor(sce$clusterSadeFeldman)

plotReducedDim(sce, "TSNE", colour_by = "clusterSadeFeldman", text_by = "clusterSadeFeldman")



ggcells(sce, mapping=aes_string(x = "CD2", y = "CD3E"), exprs_values = "logcounts") +
  geom_point() + ylab("log normalized counts CD3E") + theme_bw() 


#Define CD8 T cells and extract them from all the other subtypes
sce$isCD8T <- (counts(sce)[rowData(sce)$GeneSymbol == "CD3E",] > 0) & 
  ((counts(sce)[rowData(sce)$GeneSymbol == "CD8A",] > 0) | (counts(sce)[rowData(sce)$GeneSymbol == "CD8B",] > 0)) &
  (counts(sce)[rowData(sce)$GeneSymbol == "NCR1",] == 0) & 
  (counts(sce)[rowData(sce)$GeneSymbol == "NCAM1",] == 0) & 
  (counts(sce)[rowData(sce)$GeneSymbol == "CD4",] == 0) 
summary(sce$isCD8T)
# > summary(sce$isCD8T)
# Mode   FALSE    TRUE 
# logical   10252    5138 
plotReducedDim(sce, "TSNE", colour_by = "SNX9")

sce_CD8T <- sce[,sce$isCD8T]
counts(sce_CD8T.rds) <- NULL # remove raw counts to reduce file size
saveRDS(sce_CD8T, file = "sce_CD8T.rds")

# EASY ENTRY POINT. Start here to analyze CD8T cells directly
readRDS(sce_CD8T, file = "sce_CD8T.rds")
set.seed(123)
sce_CD8T <- runTSNE(sce_CD8T, dimred = "PCA",  perplexity = 30)

######### Plots 
pdf(file = "../plots/CD8T_SNX9_plots.pdf", width = 6.8, height= 6)
  
  sce_CD8T$UMAP1 <- reducedDim(sce_CD8T, "UMAP")[,1]
  sce_CD8T$UMAP2 <- reducedDim(sce_CD8T, "UMAP")[,2]
  sce_CD8T$PCA1 <- reducedDim(sce_CD8T, "PCA")[,1]
  sce_CD8T$PCA2 <- reducedDim(sce_CD8T, "PCA")[,2]
  sce_CD8T$TSNE1 <- reducedDim(sce_CD8T, "TSNE")[,1]
  sce_CD8T$TSNE2 <- reducedDim(sce_CD8T, "TSNE")[,2]
  
  hist(logcounts(sce_CD8T)[rowData(sce_CD8T)$GeneSymbol == "SNX9",])
  
  
  sce_CD8T$SNX9pos <- ifelse(logcounts(sce_CD8T)[rowData(sce_CD8T)$GeneSymbol == "SNX9",] > 1, "SNX9pos", "SNX9neg") 
  summary(sce_CD8T$SNX9pos)
  
  histogram_data <- data.frame("logcounts" = logcounts(sce_CD8T)[rowData(sce_CD8T)$GeneSymbol == "SNX9",], "SNX9pos" = colData(sce_CD8T)$SNX9pos ) 
  ggplot(data = histogram_data, aes(x = logcounts)) + geom_histogram(binwidth = 0.2) + geom_vline(xintercept = 1)
  ggplot(data = histogram_data, aes(x = logcounts)) + geom_histogram(binwidth = 0.2) + scale_y_log10() + labs(y = "log10 counts")+ geom_vline(xintercept = 2)
  ggplot(data = histogram_data, aes(x = logcounts)) + geom_histogram() + scale_y_log10() + facet_wrap(~SNX9pos)
  
  genes_of_interest <- c("SNX9", "PDCD1", "TCF7", "IL7R", "HAVCR2", "ENTPD1", "NR4A1", "NR4A2" , "NR4A3", "TOX", "TOX2", "CCR7", "LDHA")
  
  dat <- data.frame(as.data.frame(colData(sce_CD8T)), as.matrix(t(logcounts(sce_CD8T[rowData(sce_CD8T)$GeneSymbol %in% genes_of_interest,]))))
  head(dat, n= 20)
  
  str(dat)
  write.table(dat, file= "../lists/SNX9_cells_genesofinterest.txt", sep = "\t", quote = FALSE, row.names = FALSE)
  
  ## Plots from Figures
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = SNX9, size = SNX9)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, SNX9pos == "SNX9+"), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
  #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("SNX9 in CD8+ T cells") 
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = TCF7, size = TCF7)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, TCF7 > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
  #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("TCF7 in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = TOX, size = TOX)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, TOX > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
    #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("TOX in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = CCR7, size = CCR7)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, CCR7 > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
    #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("CCR7 in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = LDHA, size = LDHA)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, LDHA > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
    #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("LDHA in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = PDCD1, size = PDCD1)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, PDCD1 > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
    #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("PDCD1 in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = TSNE1, y = TSNE2, col = HAVCR2, size = HAVCR2)) +
    geom_point(alpha = 0.5) + 
    geom_point(data = subset(dat, HAVCR2 > 0), alpha = 0.5) + 
    theme_bw(base_size = 18) + 
    scale_color_viridis_c(option = "inferno") +
    #  scale_color_gradient( low="blue",high="red", space ="Lab" ) +
    ggtitle("HAVCR2 in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = SNX9, y = TCF7)) +
    geom_jitter(height= 1) +
    theme_bw() + 
    ggtitle("SNX9 vs TCF7 in CD8+ T cells")
  print(p)
  
  p <- ggplot(dat, aes(x = SNX9, y = TOX)) +
    geom_point() +
    theme_bw() + 
    ggtitle("SNX9 vs TOX in CD8+ T cells")
  print(p)
  
  
  
  ############3
  # compare SNX9 positive vs negative cells
  for(i in 1:length(genes_of_interest)){
    gene <- genes_of_interest[i]
    dat_selected <- dat[,c("SNX9pos", gene)]
    colnames(dat_selected) <- c("SNX9pos", "gene_value")
    w <- wilcox.test(gene_value ~ SNX9pos, data = dat_selected )
    p <- ggplot(dat_selected, aes(x = SNX9pos, y = gene_value, fill = SNX9pos)) +
      geom_violin()  + stat_summary(fun=mean, geom="point", shape=18, size=2) +
      theme_bw() + ylab(paste("log(counts)", gene)) +
      scale_fill_manual(values=c("#999999", "#108080")) +
      ggtitle(paste(gene," expression in CD8+"), sprintf("wilcoxon p-value = %.2e", w$p.value))
    print(p)
  }
  

dev.off()

pdf(file = "../plots/CD8T_SNX9_plots_patients.pdf", width = 8)
  
  patientStats <- dat %>% dplyr::select(sample, patientID, response, therapy, timepoint, patient) %>%
    distinct(patientID, .keep_all = TRUE)
  
  a <- as.data.frame.matrix(table(sce_CD8T$patientID, sce_CD8T$SNX9pos))
  colnames(a) <- c("SNX9neg", "SNX9pos")
  a$patientID <- rownames(a)
  
  patientStats <- left_join(patientStats, a, by = "patientID")
  patientStats <- patientStats %>% mutate(SNX9percent = SNX9pos / SNX9neg *100)
  
  write.table(patientStats, file= "../lists/SNX9_patientStats.txt", sep = "\t", quote = FALSE, row.names = FALSE)
  
  w <- wilcox.test(data = patientStats[patientStats$timepoint == "Pre",], SNX9percent ~ response) 
  w
  p <- ggplot(patientStats[patientStats$timepoint == "Pre",], aes(x = response, y = SNX9percent)) +
    geom_boxplot() + geom_point() +
    ylab("% SNX9+ of CD8+ T cells") +
    theme_bw()  +
    ggtitle(paste("Response vs  %SNX9 of CD8+ T cells\n", sprintf("wilcoxon p-value = %.3f", w$p.value)))
  print(p)
  
  p <- ggplot(patientStats, aes(x = response, y = SNX9percent)) +
    geom_boxplot() + geom_point() + facet_wrap(~timepoint) +
    ylab("% SNX9+ of CD8+ T cells") +
    theme_bw()  +
    ggtitle(paste("Response vs  %SNX9 of CD8+ T cells"))
  print(p)
  
  #####
  # Calculate mean expression of markers for both SNX9+ and negative cells for each sample individually
  patientStats_meanExpressions <- dat %>% dplyr::select(sample, patientID, therapy, timepoint, patient, response,  SNX9pos, genes_of_interest) %>%
    group_by(patientID, SNX9pos) %>%
    mutate(meanPDCD1 = mean(PDCD1),
           meanTCF7 = mean(TCF7),
           meanCCR7 = mean(CCR7),
           meanIL7R = mean(IL7R ),
           meanHAVCR2 = mean(HAVCR2),
           meanENTPD1 = mean(ENTPD1),
           meanNR4A1 = mean(NR4A1),
           meanNR4A2 = mean(NR4A2),
           meanNR4A3 = mean(NR4A3),
           meanTOX = mean(TOX),
           meanTOX2 = mean(TOX2)) %>%
    distinct(patientID, .keep_all = TRUE)
  
  patientStat_gathered<- patientStats %>% dplyr::select(-SNX9percent) %>% gather(SNX9pos, cell_number, SNX9neg:SNX9pos)
  
  patientStats_meanExpressions <- left_join(patientStats_meanExpressions, dplyr::select(patientStat_gathered, c("patientID", "SNX9pos", "cell_number")) , by = c("patientID", "SNX9pos"))
  
  #Table used in Figures
  write.table(patientStats_meanExpressions, file= "../lists/SNX9_patientStats_meanExpression.txt", sep = "\t", quote = FALSE, row.names = FALSE)
  
  
  p <- ggplot(patientStats_meanExpressions, aes(x = SNX9pos, y = meanPDCD1)) +
    geom_boxplot() + geom_point() + 
    ylab("mean PDCD1 expression of CD8+ T cells") +
    theme_bw()  +
    ggtitle(paste("PDCD1 Expression of CD8+ T cells"))
  print(p)
  p <- ggplot(patientStats_meanExpressions, aes(x = SNX9pos, y = meanTCF7)) +
    geom_boxplot() + geom_point() + 
    ylab("mean TCF7 expression of CD8+ T cells") +
    theme_bw()  +
    ggtitle(paste("TCF7 Expression of CD8+ T cells"))
  print(p)
  
  
  correlation <- cor(as.matrix(t(logcounts(sce_CD8T[rowData(sce_CD8T)$GeneSymbol %in% c("SNX9", "PDCD1", "TCF7", "IL7R", "HAVCR2", "ENTPD1", "NR4A3", "TOX", "TOX2"),]))))
  
  cormat <- round(correlation,2)
  head(cormat)
  library(reshape2)
  melted_cormat <- melt(cormat)
  head(melted_cormat)
  ggplot(data = melted_cormat, aes(x=Var1, y=Var2, fill=value)) + 
    geom_tile() + ggtitle("correlation between gene expression in CD8+ T cells") +
    scale_fill_gradient2(low = "blue", high = "red",  
                         midpoint = 0, limit = c(-1,1), space = "Lab", 
                         name="Pearson\nCorrelation")  +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, 
                                     size = 12, hjust = 1))


dev.off()



sessionInfo()








